src/node/txorphanage.cpp

```diff
@@ -8,68 +8,201 @@
-        TxOrphanage::Usage m_total_usage{0};
+        /** Get an approximation of how much this transaction contributes to latency in EraseForBlock and EraseForPeer.
+         * The computation time is a function of the number of entries in m_orphans (thus 1 per announcement) and the
+         * number of entries in m_outpoint_to_orphan_it (thus an additional 1 for every 10 inputs). Transactions with a
+         * small number of inputs (9 or fewer) are counted as 1 to make it easier to reason about each peer's limits in
+         * terms of "normal" transactions. */
+        TxOrphanage::Count GetLatencyScore() const {
+            return 1 + (m_tx->vin.size() / 10);
+        }
     };
-    std::map<NodeId, PeerOrphanInfo> m_peer_orphanage_info;
 
-    using OrphanMap = decltype(m_orphans);
-
-    struct IteratorComparator
+    // Index by wtxid, then peer
+    struct ByWtxid {};
+    using ByWtxidView = std::tuple<Wtxid, NodeId>;
+    struct WtxidExtractor
     {
-        template<typename I>
-        bool operator()(const I& a, const I& b) const
+        using result_type = ByWtxidView;
+        result_type operator()(const Announcement& ann) const
         {
-            return a->first < b->first;
+            return ByWtxidView{ann.m_tx->GetWitnessHash(), ann.m_announcer};
         }
     };
 
-    /** Index from the parents' COutPoint into the m_orphans. Used
-     *  to remove orphan transactions from the m_orphans */
-    std::map<COutPoint, std::set<OrphanMap::iterator, IteratorComparator>> m_outpoint_to_orphan_it;
+    // Sort by peer, then by whether it is ready to reconsider, then by recency.
+    struct ByPeer {};
+    using ByPeerView = std::tuple<NodeId, bool, SequenceNumber>;
+    struct ByPeerViewExtractor {
+        using result_type = ByPeerView;
+        result_type operator()(const Announcement& ann) const
+        {
+            return ByPeerView{ann.m_announcer, ann.m_reconsider, ann.m_entry_sequence};
+        }
+    };
+
+    struct OrphanIndices final : boost::multi_index::indexed_by<
+        boost::multi_index::ordered_unique<boost::multi_index::tag<ByWtxid>, WtxidExtractor>,
+        boost::multi_index::ordered_unique<boost::multi_index::tag<ByPeer>, ByPeerViewExtractor>
+    >{};
+
+    using AnnouncementMap = boost::multi_index::multi_index_container<Announcement, OrphanIndices>;
+    template<typename Tag>
+    using Iter = typename AnnouncementMap::index<Tag>::type::iterator;
+    AnnouncementMap m_orphans;
+
+    const TxOrphanage::Count m_max_global_latency_score{DEFAULT_MAX_ORPHANAGE_LATENCY_SCORE};
+    const TxOrphanage::Usage m_reserved_usage_per_peer{DEFAULT_RESERVED_ORPHAN_WEIGHT_PER_PEER};
+
+    /** Number of unique orphans by wtxid. Less than or equal to the number of entries in m_orphans. */
+    TxOrphanage::Count m_unique_orphans{0};
+
+    /** Memory used by orphans (see Announcement::GetMemUsage()), deduplicated by wtxid. */
+    TxOrphanage::Usage m_unique_orphan_bytes{0};
```

Comment:
in d5097d6eef85eab971352cf78fb94a30c6e6f127: nit: maybe also call it `..._usage` for consistency?